### PR TITLE
implement binary protocol (minimum)

### PR DESCRIPTION
**Pull request by xorphitus**
# purpose

We want to use katsubushi with Ruby products.

But [its popular memcached client](https://github.com/petergoldstein/dalli) supports only binary protocol.

# strategy

I've implemented only GET (single key) and VERSION which are called by [Dalli](https://github.com/petergoldstein/dalli).

Binary protocol implementation is separated from app.go.
Since text protocol is mainly maintained.

## out of scope

* performance tuning for binary protocol
  * I want to try next pull request
* implementation of other commands
* implementation of multiple key GET

# etc

If you found implementation issues (naming, conding style, ...), I'll fix as I can as possible.

Goに不慣れなため、もし変な実装をしてしまっていたら細かなことでも対応いたします。

### CODE REVIEW

**Comment by fujiwara** [comment]
Thank you for your pull request.
I'm studying about binary protocol. Please wait a few days for reviewing.

File: binary_protocol.go [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+func (app *App) RespondToBinary(r io.Reader, w io.Writer) {
+	for {
+		req, err := newRequest(r)
+		if err != nil {
+			if err != io.EOF {
+				log.Warn(err)
+			}
+			return
+		}
+
```
[the remark this answers is not in the document]
Comment (xorphitus) [reply binary_protocol.go]: Is it better that place `extendDeadline` at the beginning of this loop?
I guess, it sets timeout for reading a binary request, too.

**Comment by xorphitus** [comment]
I've fixed codes for all review comments.
Please confirm them.

File: binary_protocol.go [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+func (res bResponse) Bytes() []byte {
+	extraLen := len(res.extras)
+	keyLen := len(res.key)
+	valueLen := len(res.value)
+	totalLen := headerSize + extraLen + keyLen + valueLen
+
+	keyLenBytes := make([]byte, 2)
+	binary.BigEndian.PutUint16(keyLenBytes, uint16(keyLen))
+
+	extraLenByte := byte(uint8(extraLen))
```
Comment (shogo82148) [inline binary_protocol.go]: byte is an alias of uint8. No need to cast.

File: binary_protocol.go [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+	extraLen := uint8(buf[4])
+	bodyLen := binary.BigEndian.Uint32(buf[8:12])
+
+	if bodyLen < uint32(keyLen)+uint32(extraLen) {
+		return nil, fmt.Errorf("total body %d is too small. key length: %d, extra length %d", bodyLen, keyLen, extraLen)
+	}
+
+	bodyBuf := make([]byte, bodyLen)
+	n2, e2 := io.ReadFull(r, bodyBuf)
+	if uint32(n2) < bodyLen {
```
Comment (shogo82148) [inline binary_protocol.go]: io.ReadFull returns an error in this case.
So, no need to check n2.

File: binary_protocol.go [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+	data[16] = res.cas[0]
+	data[17] = res.cas[1]
+	data[18] = res.cas[2]
+	data[19] = res.cas[3]
+	data[20] = res.cas[4]
+	data[21] = res.cas[5]
+	data[22] = res.cas[6]
+	data[23] = res.cas[7]
+
+	for i := 0; i < extraLen; i++ {
```
Comment (shogo82148) [inline binary_protocol.go]: I think ´copy(data[headerSize:], res.extras)´ is better.

File: binary_protocol.go [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+}
+
+// BytesToCmd converts byte array to a MemdBCmd and returns it.
+func (app *App) BytesToBinaryCmd(req bRequest) (cmd MemdCmd, err error) {
+	switch req.opcode {
+	case opcodeGet:
+		atomic.AddInt64(&(app.cmdGet), 1)
+		cmd = &MemdBCmdGet{
+			Name:   "GET",
+			Keys:   strings.Fields(req.key),
```
[the remark this answers is not in the document]
Comment (xorphitus) [reply binary_protocol.go]: Thank you for reviewing!

Sorry, it's redundant as you said.
I'll remove it.

I also try to fix field name `Keys` to `Key`.

**Pull request merged**